File: src/net/merayen/elastic/ui/objects/top/viewbar/ViewBar.java
package net.merayen.elastic.ui.objects.top.viewbar;

import net.merayen.elastic.ui.Draw;
import net.merayen.elastic.ui.UIObject;
import net.merayen.elastic.ui.event.UIEvent;
import net.merayen.elastic.ui.objects.components.autolayout.AutoLayout;
import net.merayen.elastic.ui.objects.components.autolayout.LayoutMethods;
import net.merayen.elastic.ui.objects.components.dragdrop.TargetItem;
import net.merayen.elastic.ui.objects.node.EditNodeMouseCarryItem;
import net.merayen.elastic.ui.objects.node.INodeEditable;
import net.merayen.elastic.ui.objects.node.UINode;
import net.merayen.elastic.ui.objects.top.megamenu.MegaMenu;
import net.merayen.elastic.ui.objects.top.mouse.MouseCarryItem;
import net.merayen.elastic.ui.objects.top.views.View;
import net.merayen.elastic.ui.objects.top.views.editview.EditNodeView;

public class ViewBar extends AutoLayout {
	public ViewBar() {
		super(new LayoutMethods.HorizontalBox(2, 100000));
	}

	public float width;
	private final float height = 20;
	private final MegaMenu menu = new MegaMenu();
	protected final UIObject content = new UIObject();
	private TargetItem targetItem;
	private boolean interested;

	@Override
	public void onInit() {
		add(menu);
		add(content);

		targetItem = new TargetItem(this) {
			@Override
			public void onDrop(MouseCarryItem item) {
				if(item instanceof EditNodeMouseCarryItem) {
					UINode node = ((EditNodeMouseCarryItem)item).getNode();

					if(node instanceof INodeEditable) {
						EditNodeView editNodeView = getSearch().parentByType(View.class).swap(EditNodeView.class);
						editNodeView.editNode((INodeEditable) node);
					}
				}
			}

			@Override
			public void onHover(MouseCarryItem item) {}

			@Override
			public void onBlurInterest() {
				interested = false;
			}

			@Override
			public void onInterest(MouseCarryItem item) {
				interested = true;
			}

			@Override
			public void onBlur() {}
		};
	}

	@Override
	public void onDraw(Draw draw) {
		draw.setColor(100, 100, 100);
		draw.fillRect(0, 0, width, height);
		draw.setColor(0, 0, 0);
		draw.setStroke(1);
		draw.line(0, height, width, height);

		if(interested) {
			draw.setColor(255,0,255);
			draw.fillRect(0, 0, width, height);
		}
	}

	@Override
	public void onEvent(UIEvent event) {
		super.onEvent(event);
		targetItem.handle(event);
	}
}

File: src/main/kotlin/net/merayen/elastic/ui/objects/top/viewbar/ViewBar.kt
package net.merayen.elastic.ui.objects.top.viewbar

import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.FlexibleDimension
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.event.UIEvent
import net.merayen.elastic.ui.objects.components.autolayout.AutoLayout
import net.merayen.elastic.ui.objects.components.autolayout.LayoutMethods
import net.merayen.elastic.ui.objects.components.dragdrop.TargetItem
import net.merayen.elastic.ui.objects.node.EditNodeMouseCarryItem
import net.merayen.elastic.ui.objects.node.INodeEditable
import net.merayen.elastic.ui.objects.top.mouse.MouseCarryItem
import net.merayen.elastic.ui.objects.top.views.View
import net.merayen.elastic.ui.objects.top.views.editview.EditNodeView
import kotlin.reflect.KClass

open class ViewBar(private val viewClass: KClass<out View>) : AutoLayout<LayoutMethods.HorizontalBox>(LayoutMethods.HorizontalBox(2f, 100000f)), FlexibleDimension {
	override var layoutWidth = 0f
	override var layoutHeight = 40f

	private val menu = ViewSelector(object : ViewSelector.Handler {
		override fun onSelect(cls: KClass<out View>) {
			search.parentByType(View::class.java)!!.swap(cls)
		}
	})

	protected val content = UIObject()
	private var targetItem: TargetItem? = null
	private var interested: Boolean = false

	override fun onInit() {
		menu.setViewClass(viewClass)
		add(menu)
		add(content)

		targetItem = object : TargetItem(this) {
			override fun onDrop(item: MouseCarryItem) {
				if (item is EditNodeMouseCarryItem) {
					val node = item.node

					if (node is INodeEditable) {
						val editNodeView = search.parentByType(View::class.java)!!.swap(EditNodeView::class)
						editNodeView.editNode(node as INodeEditable)
					}
				}
			}

			override fun onHover(item: MouseCarryItem) {}

			override fun onBlurInterest() {
				interested = false
			}

			override fun onInterest(item: MouseCarryItem) {
				interested = true
			}

			override fun onBlur() {}
		}
	}

	override fun onDraw(draw: Draw) {
		draw.setColor(100, 100, 100)
		draw.fillRect(0f, 0f, layoutWidth, layoutHeight)
		draw.setColor(0, 0, 0)
		draw.setStroke(1f)
		draw.line(0f, layoutHeight, layoutWidth, layoutHeight)

		if (interested) {
			draw.setColor(255, 0, 255)
			draw.fillRect(0f, 0f, layoutWidth, layoutHeight)
		}
	}

	override fun onEvent(event: UIEvent) {
		super.onEvent(event)
		targetItem!!.handle(event)
	}
}